### ISSUE TITLE
Improve error messaging when browser is not found

### ISSUE THREAD

**Issue by t89s** [webhintio/hint#2768]
# 🐛 Bug report

## Description

Hello,

I am having a problem using hint.

The installation was done according to the readme

```
npm install -g --engine-strict hint
npm create hintrc
hint https://example.com

```
But it does not work
 
 
When I run hint without the debug here is the return
  
  
```
 npx hint https://example.com 
   ┌───────────────────────────────────────────────────────────────┐
   │                                                               │
   │             Couldn't find any valid configuration             │
   │                                                               │
   │         Running hint with the default configuration.          │
   │                                                               │
   │   Learn more about how to create your own configuration at:   │
   │                                                               │
   │              https://webhint.io/docs/user-guide/              │
   │                                                               │
   └───────────────────────────────────────────────────────────────┘

✖ 
 
```
 
 
I could see in the logs this
 
 
```
 hint:engine Initializing hint engine +0ms
  hint:cli:analyze Failed to analyze: undefined +0ms
  hint:cli:analyze Error: No installation found
  hint:cli:analyze     at Object.exports.getInstallationPath (/home/security/node_modules/@hint/utils/dist/src/chromium-finder.js:246:15)
  hint:cli:analyze     at PuppeteerConnector.toPuppeteerOptions (/home/security/node_modules/@hint/connector-puppeteer/dist/src/connector.js:33:21)
  hint:cli:analyze     at new PuppeteerConnector (/home/security/node_modules/@hint/connector-puppeteer/dist/src/connector.js:26:30)
  hint:cli:analyze     at new Engine (/home/security/node_modules/hint/dist/src/lib/engine.js:33:26)
  hint:cli:analyze     at Analyzer.analyze (/home/security/node_modules/hint/dist/src/lib/analyzer.js:140:28)
  hint:cli:analyze     at exports.default (/home/security/node_modules/hint/dist/src/lib/cli/analyze.js:256:23)
  hint:cli:analyze     at processTicksAndRejections (internal/process/task_queues.js:85:5)
  hint:cli:analyze     at async /home/security/node_modules/hint/dist/src/lib/cli/actions.js:7:20
  hint:cli:analyze     at async Object.exports.execute (/home/security/node_modules/hint/dist/src/lib/cli.js:38:23)
  hint:cli:analyze     at async run (/home/security/node_modules/hint/dist/src/bin/hint.js:54:24) +0ms
  hint:cli:analyze Total runtime: 52ms +1ms
```

I had it installed a few weeks ago in version 5.0.2 but today on a new install it does not work anymore.

Is there a problem of dependence / module?
```
wget -qO - https://deb.nodesource.com/setup_12.x | bash -
apt-get install gcc g++ make
apt-get install -y nodejs
npm install -g node-gyp node-pre-gyp
npm install -g --engine-strict hint
npm create hintrc
hint https://example.com

```

## Environment

* __`hint` version: v5.1.1 __ 
* __`Node.js` version: v12.7.0 __ 
* __`OS`: Debian GNU/Linux 9 \n \l __ 
* __URL for which webhint failed: https://example.com__ 

## `webhint` configuration

<details>
<summary>webhint’s configuration</summary>

```js
{
    "connector": {
        "name": "puppeteer",
        "options": {
            "waitUntil": "networkidle2"
        }
    },
    "hintsTimeout": 120000,
    "hints": {
        "amp-validator": "off",
        "apple-touch-icons": "warning",
        "button-type": "warning",
        "content-type": "warning",
        "disown-opener": "warning",
        "highest-available-document-mode": "warning",
        "html-checker": "warning",
        "http-cache": "warning",
        "http-compression": "warning",
        "image-optimization-cloudinary": "off",
        "manifest-app-name": "warning",
        "manifest-exists": "warning",
        "manifest-file-extension": "warning",
        "manifest-is-valid": "warning",
        "meta-charset-utf-8": "warning",
        "meta-viewport": "warning",
        "minified-js": "warning",
        "no-bom": "warning",
        "no-broken-links": "warning",
        "no-disallowed-headers": "warning",
        "no-friendly-error-pages": "warning",
        "no-html-only-headers": "warning",
        "no-http-redirects": "warning",
        "no-p3p": "warning",
        "no-protocol-relative-urls": "warning",
        "no-vulnerable-javascript-libraries": "warning",
        "performance-budget": "warning",
        "sri": "warning",
        "ssllabs": "off",
        "strict-transport-security": "warning",
        "stylesheet-limits": "warning",
        "validate-set-cookie-header": "warning",
        "x-content-type-options": "warning"
    }
}

```

</details>

## Debug output

Please include the raw output generated by `webhint`, or if possible,
the raw output from running `webhint` in debug mode: `hint --debug <URL>`.

<details>

```
  hint:home:security:node_modules:@hint:utils:dist:src:app-insights Application Insight disabled +0ms
  hint:home:security:node_modules:@hint:utils:dist:src:network:as-uri Adding valid target: https://example.com/ +0ms
  hint:config:config-validator Validating configuration +0ms
  hint:config Composing configuration from extends +0ms
  hint:config overriding config settings with values provided via CLI +0ms
  hint:config Initializing ignored urls +6ms
  hint:utils:resource-loader Searching puppeteer… +0ms
  hint:utils:resource-loader @hint/connector-puppeteer is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader puppeteer found in @hint/connector-puppeteer +81ms
  hint:utils:resource-loader Searching apple-touch-icons… +1ms
  hint:utils:resource-loader @hint/hint-apple-touch-icons is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader apple-touch-icons found in @hint/hint-apple-touch-icons +6ms
  hint:utils:resource-loader Searching button-type… +1ms
  hint:utils:resource-loader @hint/hint-button-type is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader button-type found in @hint/hint-button-type +2ms
  hint:utils:resource-loader Searching content-type… +1ms
  hint:utils:resource-loader @hint/hint-content-type is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader content-type found in @hint/hint-content-type +5ms
  hint:utils:resource-loader Searching disown-opener… +1ms
  hint:utils:resource-loader @hint/hint-disown-opener is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader disown-opener found in @hint/hint-disown-opener +2ms
  hint:utils:resource-loader Searching highest-available-document-mode… +1ms
  hint:utils:resource-loader @hint/hint-highest-available-document-mode is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader highest-available-document-mode found in @hint/hint-highest-available-document-mode +2ms
  hint:utils:resource-loader Searching html-checker… +1ms
  hint:utils:resource-loader @hint/hint-html-checker is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader html-checker found in @hint/hint-html-checker +2ms
  hint:utils:resource-loader Searching http-cache… +1ms
  hint:utils:resource-loader @hint/hint-http-cache is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader http-cache found in @hint/hint-http-cache +2ms
  hint:utils:resource-loader Searching http-compression… +1ms
  hint:utils:resource-loader @hint/hint-http-compression is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader http-compression found in @hint/hint-http-compression +2ms
  hint:utils:resource-loader Searching manifest-app-name… +1ms
  hint:utils:resource-loader @hint/hint-manifest-app-name is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader manifest-app-name found in @hint/hint-manifest-app-name +2ms
  hint:utils:resource-loader Searching manifest-exists… +0ms
  hint:utils:resource-loader @hint/hint-manifest-exists is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader manifest-exists found in @hint/hint-manifest-exists +1ms
  hint:utils:resource-loader Searching manifest-file-extension… +1ms
  hint:utils:resource-loader @hint/hint-manifest-file-extension is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader manifest-file-extension found in @hint/hint-manifest-file-extension +2ms
  hint:utils:resource-loader Searching manifest-is-valid… +0ms
  hint:utils:resource-loader @hint/hint-manifest-is-valid is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader manifest-is-valid found in @hint/hint-manifest-is-valid +3ms
  hint:utils:resource-loader Searching meta-charset-utf-8… +1ms
  hint:utils:resource-loader @hint/hint-meta-charset-utf-8 is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader meta-charset-utf-8 found in @hint/hint-meta-charset-utf-8 +2ms
  hint:utils:resource-loader Searching meta-viewport… +0ms
  hint:utils:resource-loader @hint/hint-meta-viewport is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader meta-viewport found in @hint/hint-meta-viewport +2ms
  hint:utils:resource-loader Searching minified-js… +1ms
  hint:utils:resource-loader @hint/hint-minified-js is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader minified-js found in @hint/hint-minified-js +2ms
  hint:utils:resource-loader Searching no-bom… +0ms
  hint:utils:resource-loader @hint/hint-no-bom is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader no-bom found in @hint/hint-no-bom +2ms
  hint:utils:resource-loader Searching no-broken-links… +1ms
  hint:utils:resource-loader @hint/hint-no-broken-links is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader no-broken-links found in @hint/hint-no-broken-links +2ms
  hint:utils:resource-loader Searching no-disallowed-headers… +0ms
  hint:utils:resource-loader @hint/hint-no-disallowed-headers is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader no-disallowed-headers found in @hint/hint-no-disallowed-headers +2ms
  hint:utils:resource-loader Searching no-friendly-error-pages… +0ms
  hint:utils:resource-loader @hint/hint-no-friendly-error-pages is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader no-friendly-error-pages found in @hint/hint-no-friendly-error-pages +2ms
  hint:utils:resource-loader Searching no-html-only-headers… +1ms
  hint:utils:resource-loader @hint/hint-no-html-only-headers is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader no-html-only-headers found in @hint/hint-no-html-only-headers +2ms
  hint:utils:resource-loader Searching no-http-redirects… +0ms
  hint:utils:resource-loader @hint/hint-no-http-redirects is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader no-http-redirects found in @hint/hint-no-http-redirects +1ms
  hint:utils:resource-loader Searching no-p3p… +1ms
  hint:utils:resource-loader @hint/hint-no-p3p is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader no-p3p found in @hint/hint-no-p3p +24ms
  hint:utils:resource-loader Searching no-protocol-relative-urls… +1ms
  hint:utils:resource-loader @hint/hint-no-protocol-relative-urls is not a JSON file, trying to load it normally +6ms
  hint:utils:resource-loader no-protocol-relative-urls found in @hint/hint-no-protocol-relative-urls +2ms
  hint:utils:resource-loader Searching no-vulnerable-javascript-libraries… +5ms
  hint:utils:resource-loader @hint/hint-no-vulnerable-javascript-libraries is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader no-vulnerable-javascript-libraries found in @hint/hint-no-vulnerable-javascript-libraries +12ms
  hint:utils:resource-loader Searching performance-budget… +1ms
  hint:utils:resource-loader @hint/hint-performance-budget is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader performance-budget found in @hint/hint-performance-budget +3ms
  hint:utils:resource-loader Searching sri… +0ms
  hint:utils:resource-loader @hint/hint-sri is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader sri found in @hint/hint-sri +2ms
  hint:utils:resource-loader Searching strict-transport-security… +1ms
  hint:utils:resource-loader @hint/hint-strict-transport-security is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader strict-transport-security found in @hint/hint-strict-transport-security +2ms
  hint:utils:resource-loader Searching stylesheet-limits… +0ms
  hint:utils:resource-loader @hint/hint-stylesheet-limits is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader stylesheet-limits found in @hint/hint-stylesheet-limits +1ms
  hint:utils:resource-loader Searching validate-set-cookie-header… +9ms
  hint:utils:resource-loader @hint/hint-validate-set-cookie-header is not a JSON file, trying to load it normally +0ms
  hint:utils:resource-loader validate-set-cookie-header found in @hint/hint-validate-set-cookie-header +2ms
  hint:utils:resource-loader Searching x-content-type-options… +0ms
  hint:utils:resource-loader @hint/hint-x-content-type-options is not a JSON file, trying to load it normally +1ms
  hint:utils:resource-loader x-content-type-options found in @hint/hint-x-content-type-options +2ms
  hint:config Validating puppeteer connector +230ms
  hint:utils:resource-loader Searching puppeteer… +5ms
  hint:config Connector schema: +3ms
  hint:config {
  hint:config   additionalProperties: false,
  hint:config   definitions: {
  hint:config     fieldInput: { properties: [Object], required: [Array], type: 'object' },
  hint:config     submitInput: { properties: [Object], required: [Array], type: 'object' }
  hint:config   },
  hint:config   properties: {
  hint:config     auth: {
  hint:config       additionalProperties: false,
  hint:config       properties: [Object],
  hint:config       required: [Array],
  hint:config       type: 'object'
  hint:config     },
  hint:config     browser: { enum: [Array], type: 'string' },
  hint:config     detached: { type: 'boolean' },
  hint:config     headless: { type: 'boolean' },
  hint:config     ignoreHTTPSErrors: { type: 'boolean' },
  hint:config     puppeteerOptions: { type: 'object' },
  hint:config     waitUntil: { enum: [Array], type: 'string' }
  hint:config   }
  hint:config } +0ms
  hint:config User configuration: +5ms
  hint:config { waitUntil: 'networkidle2' } +0ms
  hint:utils:resource-loader Searching apple-touch-icons… +31ms
  hint:config:config-hints Validating hint apple-touch-icons +0ms
  hint:utils:resource-loader Searching button-type… +0ms
  hint:config:config-hints Validating hint button-type +0ms
  hint:utils:resource-loader Searching content-type… +0ms
  hint:config:config-hints Validating hint content-type +0ms
  hint:utils:resource-loader Searching disown-opener… +1ms
  hint:config:config-hints Validating hint disown-opener +1ms
  hint:utils:resource-loader Searching highest-available-document-mode… +0ms
  hint:config:config-hints Validating hint highest-available-document-mode +0ms
  hint:utils:resource-loader Searching html-checker… +0ms
  hint:config:config-hints Validating hint html-checker +0ms
  hint:utils:resource-loader Searching http-cache… +0ms
  hint:config:config-hints Validating hint http-cache +1ms
  hint:utils:resource-loader Searching http-compression… +1ms
  hint:config:config-hints Validating hint http-compression +0ms
  hint:utils:resource-loader Searching manifest-app-name… +0ms
  hint:config:config-hints Validating hint manifest-app-name +0ms
  hint:utils:resource-loader Searching manifest-exists… +0ms
  hint:config:config-hints Validating hint manifest-exists +0ms
  hint:utils:resource-loader Searching manifest-file-extension… +1ms
  hint:config:config-hints Validating hint manifest-file-extension +1ms
  hint:utils:resource-loader Searching manifest-is-valid… +0ms
  hint:config:config-hints Validating hint manifest-is-valid +0ms
  hint:utils:resource-loader Searching meta-charset-utf-8… +0ms
  hint:config:config-hints Validating hint meta-charset-utf-8 +0ms
  hint:utils:resource-loader Searching meta-viewport… +0ms
  hint:config:config-hints Validating hint meta-viewport +0ms
  hint:utils:resource-loader Searching minified-js… +1ms
  hint:config:config-hints Validating hint minified-js +1ms
  hint:utils:resource-loader Searching no-bom… +0ms
  hint:config:config-hints Validating hint no-bom +0ms
  hint:utils:resource-loader Searching no-broken-links… +0ms
  hint:config:config-hints Validating hint no-broken-links +0ms
  hint:utils:resource-loader Searching no-disallowed-headers… +0ms
  hint:config:config-hints Validating hint no-disallowed-headers +1ms
  hint:utils:resource-loader Searching no-friendly-error-pages… +1ms
  hint:config:config-hints Validating hint no-friendly-error-pages +0ms
  hint:utils:resource-loader Searching no-html-only-headers… +0ms
  hint:config:config-hints Validating hint no-html-only-headers +0ms
  hint:utils:resource-loader Searching no-http-redirects… +0ms
  hint:config:config-hints Validating hint no-http-redirects +1ms
  hint:utils:resource-loader Searching no-p3p… +1ms
  hint:config:config-hints Validating hint no-p3p +0ms
  hint:utils:resource-loader Searching no-protocol-relative-urls… +0ms
  hint:config:config-hints Validating hint no-protocol-relative-urls +1ms
  hint:utils:resource-loader Searching no-vulnerable-javascript-libraries… +1ms
  hint:config:config-hints Validating hint no-vulnerable-javascript-libraries +0ms
  hint:utils:resource-loader Searching performance-budget… +0ms
  hint:config:config-hints Validating hint performance-budget +0ms
  hint:utils:resource-loader Searching sri… +1ms
  hint:config:config-hints Validating hint sri +1ms
  hint:utils:resource-loader Searching strict-transport-security… +0ms
  hint:config:config-hints Validating hint strict-transport-security +0ms
  hint:utils:resource-loader Searching stylesheet-limits… +0ms
  hint:config:config-hints Validating hint stylesheet-limits +0ms
  hint:utils:resource-loader Searching validate-set-cookie-header… +0ms
  hint:config:config-hints Validating hint validate-set-cookie-header +0ms
  hint:utils:resource-loader Searching x-content-type-options… +1ms
  hint:config:config-hints Validating hint x-content-type-options +1ms
  hint:home:security:node_modules:@hint:utils:dist:src:app-insights Sending event "analyze" to Application Insights with value {"connector":{"name":"puppeteer","options":{"waitUntil":"networkidle2"}},"hintsTimeout":120000,"hints":{"amp-validator":"off","apple-touch-icons":"warning","button-type":"warning","content-type":"warning","disown-opener":"warning","highest-available-document-mode":"warning","html-checker":"warning","http-cache":"warning","http-compression":"warning","image-optimization-cloudinary":"off","manifest-app-name":"warning","manifest-exists":"warning","manifest-file-extension":"warning","manifest-is-valid":"warning","meta-charset-utf-8":"warning","meta-viewport":"warning","minified-js":"warning","no-bom":"warning","no-broken-links":"warning","no-disallowed-headers":"warning","no-friendly-error-pages":"warning","no-html-only-headers":"warning","no-http-redirects":"warning","no-p3p":"warning","no-protocol-relative-urls":"warning","no-vulnerable-javascript-libraries":"warning","performance-budget":"warning","sri":"warning","ssllabs":"off","strict-transport-security":"warning","stylesheet-limits":"warning","validate-set-cookie-header":"warning","x-content-type-options":"warning"},"language":"en-US"} +971ms
  hint:home:security:node_modules:@hint:utils:dist:src:app-insights Application Insights is not enabled. +0ms
  hint:engine Initializing hint engine +0ms
  hint:cli:analyze Failed to analyze: undefined +0ms
  hint:cli:analyze Error: No installation found
  hint:cli:analyze     at Object.exports.getInstallationPath (/home/security/node_modules/@hint/utils/dist/src/chromium-finder.js:246:15)
  hint:cli:analyze     at PuppeteerConnector.toPuppeteerOptions (/home/security/node_modules/@hint/connector-puppeteer/dist/src/connector.js:33:21)
  hint:cli:analyze     at new PuppeteerConnector (/home/security/node_modules/@hint/connector-puppeteer/dist/src/connector.js:26:30)
  hint:cli:analyze     at new Engine (/home/security/node_modules/hint/dist/src/lib/engine.js:33:26)
  hint:cli:analyze     at Analyzer.analyze (/home/security/node_modules/hint/dist/src/lib/analyzer.js:140:28)
  hint:cli:analyze     at exports.default (/home/security/node_modules/hint/dist/src/lib/cli/analyze.js:256:23)
  hint:cli:analyze     at processTicksAndRejections (internal/process/task_queues.js:85:5)
  hint:cli:analyze     at async /home/security/node_modules/hint/dist/src/lib/cli/actions.js:7:20
  hint:cli:analyze     at async Object.exports.execute (/home/security/node_modules/hint/dist/src/lib/cli.js:38:23)
  hint:cli:analyze     at async run (/home/security/node_modules/hint/dist/src/bin/hint.js:54:24) +0ms
  hint:cli:analyze Total runtime: 44ms +0ms
```

</details>

**Comment by molant** [comment]
Errors seem to suggest that a chrome/chromium installation isn't found.
Can you please paste the path to your chrome binary and tell us what version it is?
IIRC there's a property in hintrc that can be used to specify the path for the puppeteer connector. I'm on my phone with very limited connectivity so I can't give you the right link but it should be in the docs and hopefully unlock you while someone can take a deeper look.


________________________________

**Comment by antross** [comment]
Thanks for the report @totof68!

Looking at the debug logs, I see webhint can't find a Chromium installation to use for analysis. We recently updated our default configuration, `configuration-web-recommended`, to use `connector-puppeteer` which needs a local Chromium instance to work (can be Chromium or Chrome).

If installing a Chromium instance isn't an option for you, you can switch back to the previous behavior by updating your `.hintrc` to use the `jsdom` connector instead (which was the old default).

```json
{
    "connector": "jsdom",
    "extends": ["web-recommended"]
}
```

**Comment by antross** [comment]
@molant This makes me a bit concerned about relying on a local Chromium install for our default configuration to work. While most people already have a local install, not everyone will.

Perhaps we should consider using `puppeteer` instead of `puppeteer-core` which includes a bundled Chromium instance. This would give us something we can use in the absence of finding a local install while still allowing users to configure `connector-puppeteer` to load the Chromium instance of their choosing. Thoughts?

**Comment by antross** [comment]
> IIRC there's a property in hintrc that can be used to specify the path for the puppeteer connector.

Per @molant's reference I see this is currently missing from the [documentation for `connector-puppeteer`](https://webhint.io/docs/user-guide/connectors/connector-puppeteer/), but you can indeed specify a path to a local Chromium executable in your `.hintrc` if you have one that webhint isn't finding:

```json
{
    "connector": {
        "name": "puppeteer",
        "options": {
            "puppeteerOptions": {
                "executablePath": "/path/to/Chrome"
            }
        }
    },
    "extends": ["web-recommended"]
}
```

**Comment by molant** [comment]
Using puppeteer instead of core add more than 100MB to the install process. Don't think this should be the default but maybe there's something we can do to pull it if it's not found after asking the user.
In any case maybe we can bubble up the error message more?


________________________________

**Comment by totof68** [comment]
Hello

Thank you for your answers !
Chrome / Chromium is not present.
I am doing this install for a client on a server that does not have a GUI.

By changing the connector with jsdom it works perfectly!

**Comment by molant** [comment]
This was fixed via 42b40b58faf52d794d6cadd4e40de8fbfe8b93fd
